v2/internal/frontend/desktop/windows/go-webview2/pkg/combridge/bridge.go

```diff
@@ -0,0 +1,221 @@
+// to the specified go interface.
+func New[T IUnknown](obj T) *ComObject[T] {
+	cObj := new(
+		ifceDef[T]{obj},
+	)
+	return newComObject[T](cObj)
+}
+
+// New returns a new ComObject which implements the specified Com Interface, com calls will be redirected
+// to the specified go interface.
+func New2[T IUnknown, T2 IUnknown](obj T, obj2 T2) *ComObject[T] {
+	cObj := new(
+		ifceDef[T]{obj},
+		ifceDef[T2]{obj2},
+	)
+	return newComObject[T](cObj)
+}
+
+// New returns a new ComObject which implements the specified Com Interface, com calls will be redirected
+// to the specified go interface.
+func new(impls ...ifceImpl) *comObject {
+	impls = append([]ifceImpl{ifceDef[IUnknown]{}}, impls...)
+
+	cObj := &comObject{
+		refCount:  1,
+		ifces:     map[string]int{},
+		ifcesImpl: make([]comInterfaceDesc, len(impls)),
+	}
+
+	for i, ifceDef := range impls {
+		vtable, err := ifceDef.ifce()
+		if err != nil {
+			panic(err)
+		}
+
+		needsImplement := false
+		for table := vtable; table != nil; table = table.Parent {
+			guid := table.ComGUID
+			if i, found := cObj.ifces[guid]; found {
+				// This Interface is already implemented
+				if guid == iUnknownGUID {
+					// IUnknown is a special interface and never has an user specific implementation
+				} else if cObj.ifcesImpl[i].impl != ifceDef.impl() {
+					panic(fmt.Sprintf("Interface '%s' is already implemented by another object", table.Name))
+				}
+
+				break
+			}
+
+			needsImplement = true
+			cObj.ifces[guid] = i
+		}
+
+		if !needsImplement {
+			continue
+		}
+
+		ifceP, ifcePSlice := allocUintptrObject(1)
+		ifcePSlice[0] = vtable.ComVTable
+		cObj.ifcesImpl[i] = comInterfaceDesc{ifceP, ifceDef.impl()}
+	}
+
+	comIfcePointersL.Lock()
+	for _, ifceImpl := range cObj.ifcesImpl {
+		comIfcePointers[ifceImpl.ref] = cObj
+	}
+	comIfcePointersL.Unlock()
+
+	return cObj
+}
+
+func newComObject[T IUnknown](comObj *comObject) *ComObject[T] {
+	c := &ComObject[T]{obj: comObj}
+	runtime.SetFinalizer(c, func(obj *ComObject[T]) { obj.Close() })
+	return c
+}
+
+// ComObject describes an exported go instance to be used as a ComObject which implements
+// the specified Interface.
+type ComObject[T IUnknown] struct {
+	obj    *comObject
+	closed int32
+}
+
+// Ref returns the native uintptr that points to the ComObject that is an interface pointer to T.
+// This can be used in native calls. If the object has been closed this function will panic.
+func (o *ComObject[T]) Ref() uintptr {
+	if atomic.LoadInt32(&o.closed) != 0 {
+		panic("ComObject has been released")
+	}
+	return o.obj.queryInterface(guidOf[T](), false)
+}
+
+// Close releases the native com object from the go side. It will only be destroyed if the ref counter
+// reaches zero.
+// After closing `Ref()` will panic.
+func (o *ComObject[T]) Close() error {
+	if atomic.CompareAndSwapInt32(&o.closed, 0, 1) {
+		runtime.SetFinalizer(o, nil)
+		o.obj.release()
```

Comment:
This should be done in a new goroutine because it acquires locks internally, which might block the finalizer goroutine for a longer time.
Will add a new commit with a fix for that.

Reply:
Just pushed the fix for this one.